node/src/stratum.rs

```diff
@@ -0,0 +1,1637 @@
+                let sibling_hex = hex::encode(sibling_node);
+                merkel_branches.push(sibling_hex);
+            }
+        }
+        log::info!(
+            "Merkel branches for the given template's coinbase are respectively - {:?}",
+            merkel_branches
+        );
+        //stratum accepts the prev block hash to be in little endian instead of big endian
+        //therefore byte by byte reversal is required here
+        let mut prev_block_hash = notified_template.previousblockhash.as_str();
+        let prev_block_hash_little_endian = to_little_endian(prev_block_hash);
+        log::info!(
+            "Converting the prev block hash to little endian done -- {:?}",
+            prev_block_hash_little_endian
+        );
+        let bitcoin_block_version = notified_template.version;
+        let bits = notified_template.bits;
+        let time = notified_template.curtime;
+        Ok(JobNotification {
+            job_id: new_job_id.to_string(),
+            prevhash: prev_block_hash_little_endian,
+            coinbase1: coinbase_1,
+            coinbase2: coinbase_2,
+            merkle_branches: merkel_branches,
+            //converting the i32 version to hex string
+            version: hex::encode(bitcoin_block_version.to_be_bytes()),
+            //String is acceptable
+            nbits: bits,
+            //ntime is to be hex encoded
+            ntime: hex::encode(time.to_be_bytes()),
+            clean_jobs: clean_job,
+        })
+    }
+    ///Will run a notifier that will serve the purpose of mining.notify to provide the downstream nodes with valid jobs on the basis
+    /// of the latest block template recieved/present via the template_receiver
+    pub async fn run_notifier(
+        &mut self,
+        downstream_connection_map: Arc<Mutex<ConnectionMapping>>,
+        latest_template_arc: &mut Arc<Mutex<BlockTemplate>>,
+        latest_template_merkel_branch_arc: &mut Arc<Mutex<Vec<Vec<u8>>>>,
+    ) -> Result<(), StratumErrors> {
+        log::info!("Notifier task has  started");
+        while let Some(notification_command) = self.notification_receiver.recv().await {
+            match notification_command {
+                //Whenever a new template is received it is broadcasted across all the downstream nodes connected
+                NotifyCmd::SendToAll {
+                    template,
+                    merkel_branch_coinbase,
+                } => {
+                    log::info!("Received new template to broadcast to all clients");
+                    let template_ref = template.clone();
+                    //We will receive the template from the IPC channel and construct a valid job
+                    //from the provided template and pass onto the message_reciver in the handle connection for
+                    // downstream communication to take place.
+                    for (peer_adr, mining_job_arc) in self.job_map_arc.lock().await.iter() {
```

Comment:
This holds this lock for the entire duration of notifying all clients. Clone the job_map instead.